cookbooks/fb_storage/libraries/storage_handlers.rb

```diff
@@ -0,0 +1,585 @@
+            umount(m)
+          end
+        end
+      end
+
+      def umount_device
+        if @node['filesystem2']['by_device'][@device] &&
+            @node['filesystem2']['by_device'][@device]['mounts']
+          @node['filesystem2']['by_device'][@device]['mounts'].each do |m|
+            umount(m)
+          end
+        end
+      end
+
+      def umount(m)
+        # we may call umount on the same thing more than once depending on
+        # our path through the system, so check it's actually mounted.
+        if Pathname.new(m).mountpoint?
+          Chef::Log.info("fb_storage: Unmounting #{m}")
+          Mixlib::ShellOut.new("/bin/umount #{m}").run_command.error!
+        end
+      end
+
+      def partition_device_name(num)
+        FB::Storage.partition_device_name(@device, num)
+      end
+
+      def existing_partitions
+        @existing_partitions_cache ||=
+          @node['filesystem2']['by_device'].keys.select do |x|
+            x.start_with?(@device) && x != @device
+          end
+      end
+
+      def mkfs_cmd(type)
+        case type
+        when 'xfs'
+          'mkfs -t xfs -f'
+        when 'btrfs'
+          'mkfs.btrfs -f'
+        when 'ext4', 'ext3', 'ext2'
+          "mkfs -t #{type} -F"
+        end
+      end
+
+      def default_format_options(type)
+        case type
+        when 'xfs'
+          '-i size=2048'
+        when 'btrfs'
+          '-l 16K -n 16K'
+        when 'ext4'
+          ''
+        end
+      end
+
+      class FioHandler < FB::Storage::Handler
+        def initialize(device, node)
+          super
+          @type = :fio
+          raw = device.sub('fio', 'fct')
+          num = raw[-1].tr('[a-j]', '[0-9]')
+          raw[-1] = num
+          @raw_device = raw
+        end
+
+        def self.match?(devname, _info)
+          devname.start_with?('fio')
+        end
+
+        def prep_device
+          # "format" isn't "make a filesystem" but some other magical
+          # flash prep thing
+          { 'detach' => nil,
+            'format' => '-y',
+            'attach' => nil }.each do |step, opts|
+            cmd = "/usr/bin/fio-#{step}"
+            cmd << " #{opts}" if opts
+            Chef::Log.debug(
+              "fb_storage: Running #{cmd} #{@raw_device}",
+            )
+            Mixlib::ShellOut.new("#{cmd} #{@raw_device}").
+              run_command.error!
+          end
+        end
+      end
+
+      class JbodHandler < FB::Storage::Handler
+        def initialize(device, node)
+          super
+          @type = :jbod
+          # Current AOE drivers have a bug where big partitions can take
+          # more than 10 minutes to run mkfs
+          self.mkfs_timeout = 900
+        end
+
+        # JBOD always matches as it'll work for every block device
+        def self.match?(_devname, _info)
+          true
+        end
+      end
+
+      class MdHandler < FB::Storage::Handler
```

Comment:
I don't think the nvme handler is particularly private, did we drop that for some reason?

Reply:
It currently shells out to the internal `flash_manager` to get card data. It will eventually get decoupled and opensourced as well, but it's going to take a while.